azure-kusto-ingest/azure/kusto/ingest/ingest_client.py

```diff
@@ -108,3 +125,32 @@ def ingest_from_blob(self, blob_descriptor: BlobDescriptor, ingestion_properties
         content = ingestion_blob_info_json
         queue_client = queue_service.get_queue_client(queue=random_queue.object_name, message_encode_policy=TextBase64EncodePolicy())
         queue_client.send_message(content=content)
+
+    def validate_endpoint_service_type(self):
+        if not hasattr(self, "_endpoint_service_type") or len(self._endpoint_service_type) == 0 or self._endpoint_service_type.isspace():
+            self._endpoint_service_type = self._retrieve_service_type()
```

Comment:
The exceptions thrown by this API are meaningless to the SDK user. 
Suggest you catch and wrap with a user friendly notice

Reply:
I preferred this approach to initializing to a default because it's a tradeoff between the extra checks on those 2 lines vs dirtying code that generally doesn't need to worry about any of these things. In most cases, the variables I use to keep track of whether in the past we've encountered the wrong endpoint are meaningless and therefore adds unnecessary complexity, both in the init code and while debugging.

I think we can achieve both my and your goals (simplicity both when the wrong endpoint is irrelevant, and on this particular line) by extracting this line to a function. WDYT?

Reply:
"Suggest you catch and wrap with a user friendly notice"
Are you referring to the Error type (KustoClientError), or the message?
I assume the former, because I do include a user-friendly message.

In terms of my approach, I did this to copy the paradigm in the Java SDK where it always returns to the client either a Client or Server exception. I thought this SDK was the same way and therefore the preferred way to do it, but now that I look deeper I see there are a couple that were added. I'll change the error type.

Reply:
1. I disagree, you over-complicate. Just because python allows creating class members at any point, it does not mean this is a good practice.
2. The later. This message in these exception are meaningless to the SDK user

Reply:
to be clear i refer to exceptions that might be thrown from retrieve_service_type.
The user has no clue you run .show version in this context.
Try catch it and raise a meaningful exception, with the source exception attached internally

Reply:
1. You don't see the benefit of keeping the unnecessary wrong-endpoint-detection variables out of the happy path flow's code, or you do but don't think it's worthwhile creating a class instance variable on the fly?

2. I agree that the error in retrieve_service_type isn't intuitive to the user because they wouldn't even know that getting the endpoint's version is part of the flow. But I can't think of anything more user-friendly for this issue, other than for them to just know that something went wrong in the Kusto Client or on the Kusto Server. And they are told exactly that, because the error raised is a KustoClientError or KustoServiceError.
Either way, they'll have to look into the SDK deeper or open a ticket, and the message is a sort of internal attachment that will help them debug/us address the ticket.
Can you think of a better message or error type that would actually help the SDK user to *not* have to look deeper into the SDK code/open a ticket?

Reply:
I see you resolved '2'

As for '1', in my opinion you are abusing a capacity of Python to add class members at any point.
A class which has different attributes depending on which APIs are called is prone to faults, hard to comprehend, and hard to debug. 
Additionally, Should I want to use those attributes elsewhere I'll be forced to add those checks again and again.

Keep in mind that good code needs to be readable and maintainable, and sometimes this comes at the expanse of absolute efficiency.

As a compromise, **if you need to use this more then once**, you can wrap this in a property which checks for existence and calls the function.

Reply:
I didn't change (2). The error type and error message are how I originally wrote them. I had asked if you has a better suggestion. However, per another comment I just wrote, I'm removing those lines altogether now.


Regarding (1), I see **you and @ArielYehezkely both agree and insist, so I'll change it**.

But for the record, I disagree and find the approach I wrote to be easier to debug, and more readable and maintainable. I put those variables only there because the feature is confined to only that section of code and I can't conceive of a scenario where they'd be used elsewhere.

As I mentioned, this is my first Python PR I've ever written, so I looked a lot up. And as far as I read, defining instance variables at runtime is very much considered good practice in scenarios like this by the Python community.

In fact, I think the way I did it here is common among popular Python modules. I found many... 1 example is the "http" module's server.py's _headers_buffer.